---
title: "v1_first_experiment"
author: "Giulia Moro, Izaskun Mallona, Mark D. Robinson"
format: 
  pdf: 
    toc: true
    embed-resources: true
    keep_md: true
    documentclass: article
editor_options: 
  chunk_output_type: console
---

\fontsize{8pt}{8pt}\selectfont

### Load packages, load data 

```{r,warning=FALSE, message=FALSE}

rm(list=ls())

library(SingleCellExperiment)
library(scran)
library(ggplot2)
library(dplyr)
library(scater)
library(reshape2)
library(UpSetR)
library(limma)
library(cowplot)
library(pheatmap)
library(readr)
library(plotROC)
library(here)
library(RColorBrewer)
library(ggnewscale)
library(scales)
library(rmarkdown)
library(irlba)
library(viridis)
library(gridExtra)
library(googlesheets4)
library(tools)
library(viridis)
library(scDblFinder)
library(ggrastr)
library(khroma)
library(Matrix)
library(edgeR)
library(GGally)


```

### Setting up knitr 

```{r}

knitr::opts_chunk$set(fig.width = 2.7,
               fig.height = 2.7,
               fig.align = "center",
               include = TRUE,
               dev = c("png","svg"),
               cache=TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)

options(width=80)
knitr::opts_chunk$set(echo = TRUE,
                      out.width = 40, 
                      tidy=T, 
                      tidy.opts=list(width.cutoff=60))

```

### Themes

```{r}

rock_roi_theme<-function(){
  theme(axis.text = element_text(size=5), 
        plot.title=element_text(size=7,face="bold",hjust=0.5),
        axis.title=element_text(size=7,face="bold"),
        strip.text=element_text(size=7,face="bold"),
        legend.title=element_text(size=7,face="bold"),
        legend.text=element_text(size=5),
        aspect.ratio=1,
        legend.key.size = unit(0.5, 'cm'),
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-5,-5,-5,-5),
        legend.title.align=0.5,
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())
}

rock_roi_theme_multiple_panels<-function(){
    theme(axis.text = element_text(size=5), 
        plot.title=element_text(size=7,face="bold",hjust=0.5),
        axis.title=element_text(size=7,face="bold"),
        strip.text=element_text(size=7,face="bold"),
        aspect.ratio=1,
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        legend.title=element_text(size=7,face="bold"),
        legend.text=element_text(size=7),
        legend.margin=margin(0,0,0,0),
        legend.box.margin=margin(-5,-5,-5,-5),
        legend.title.align=0.5,
        panel.border = element_rect(colour = "black", fill = NA))
}

```

### Colors 

```{r,out.width = '70%'}

myscale<-c("#66CCEE","#AA3377")

# defining colors for samples and for species

names(myscale) <- c("unmod","rock")

#plot(1:2, col=myscale, pch=19, cex=4)

```

### Load data

```{r}

exps <- c('unmod_1','unmod_3','mod_2','mod_4')
d <- list()

fpath<-'/Users/gimoro/Desktop/v1_analysis_thesis/data_first_experiment'

for (experiment in exps) {

    counts <- readMM(file.path(fpath,experiment,'matrix.mtx'))
    genes <- read.table(file.path(fpath,experiment,'features.tsv'),header = FALSE)
    cell_ids <- read.table(file.path(fpath,experiment,'barcodes.tsv'),header = FALSE)

    gene_ids <- paste0(genes$V1,'_',genes$V2)
    cell_ids<-cell_ids$V1
    
    rownames(counts) <- gene_ids
    colnames(counts) <- cell_ids
    
    d[[experiment]] <- SingleCellExperiment(counts)
    
    rowData(d[[experiment]])$gene_type<-"capture"
    k<-grepl("^ENSMUS",rownames(d[[experiment]]))
    rowData(d[[experiment]])$gene_type[k]<-"mouse_gene"
    k <- grepl("^mt",genes$V2)
    rowData(d[[experiment]])$gene_type[k]<-"mouse_mito"
    
    d[[experiment]]$sample_id<-experiment
    mainExpName(d[[experiment]])<-experiment
    rm(counts, genes, cell_ids)
}

for (i in d){
  print(paste0(unique(i$sample_id),"__",length(colnames(i))))
}

```

### Downsampling count tables 

```{r}

for (i in d){
  print(paste0(mainExpName(i),"_",sum(assay(i)))) # lowest is mod_2 with counts 33667395
}

```

### Checking colsums for each cell

```{r}

mean(colSums(assay(d[["unmod_1"]]))) 
mean(colSums(assay(d[["unmod_3"]]))) 
mean(colSums(assay(d[["mod_2"]]))) # lowest is 6390.778
mean(colSums(assay(d[["mod_4"]]))) 

proportion_unmod_1<-mean(colSums(assay(d[["mod_2"]])))/
                    mean(colSums(assay(d[["unmod_1"]])))
proportion_unmod_3<-mean(colSums(assay(d[["mod_2"]])))/
                    mean(colSums(assay(d[["unmod_3"]])))
proportion_mod_4<-mean(colSums(assay(d[["mod_2"]])))/
                  mean(colSums(assay(d[["mod_4"]])))

test_unmod_1<-downsampleMatrix(x=assay(d[["unmod_1"]]), 
                               prop=proportion_unmod_1, 
                               bycol = TRUE, 
                               sink = NULL)

test_unmod_3<-downsampleMatrix(x=assay(d[["unmod_3"]]), 
                               prop=proportion_unmod_3, 
                               bycol = TRUE, 
                               sink = NULL)

test_mod_4<-downsampleMatrix(x=assay(d[["mod_4"]]), 
                             prop=proportion_mod_4, 
                             bycol = TRUE, 
                             sink = NULL)

# checking if it is to the same mean colsum

mean(colSums(assay(d[["mod_2"]])))
mean(colSums(test_unmod_1))
mean(colSums(test_unmod_3))
mean(colSums(test_mod_4))

# generating sce objects

downsampled_d<-d

assay(downsampled_d$unmod_1)<-test_unmod_1
assay(downsampled_d$unmod_3)<-test_unmod_3
assay(downsampled_d$mod_2)<-assay(d$mod_2)
assay(downsampled_d$mod_4)<-test_mod_4

rm(d)

downsampled_d$unmod_1$sample_id<-"downsampled_unmod_1"
downsampled_d$unmod_3$sample_id<-"downsampled_unmod_3"
downsampled_d$mod_2$sample_id<-"downsampled_mod_2"
downsampled_d$mod_4$sample_id<-"downsampled_mod_4"

sce <- Reduce(cbind, downsampled_d)

table(sce$sample_id)

```

### Data QC

```{r}

rowData(sce)$gene_type %>% table

```

### QC plots

```{r,out.width = '70%'}

names(assays(sce))<-"counts"

rd <- rowData(sce)
sce <- addPerCellQCMetrics(sce,
                           subsets=list(capture=rd$gene_type=="capture",
                                        mouse_gene=rd$gene_type=="mouse_gene",
                                        mouse_mito=rd$gene_type=="mouse_mito"))

cd <- colData(sce) %>% as.data.frame 

plot1<-ggplot(cd, aes(x=factor(sample_id,
                               levels=c("downsampled_unmod_1","downsampled_unmod_3",
                                        "downsampled_mod_2","downsampled_mod_4")), 
                      y=detected)) +
        geom_violin()+  
  theme_bw()+
  rock_roi_theme_multiple_panels()+
  xlab("")+
  ylab("# genes")+ 
  geom_jitter(aes(color=sample_id),
              shape=16, 
              position=position_jitter(0.2),
              size=0.1)+
  geom_boxplot(width=0.3,
               outlier.size = 0.5,
               alpha=0.5)+
  theme(legend.position="none")+
  scale_color_manual(labels=c("downsampled_unmod_1","downsampled_unmod_3",
                              "downsampled_mod_2","downsampled_mod_4"), 
                     values=alpha(c(myscale[["unmod"]],myscale[["unmod"]],myscale[["rock"]],myscale[["rock"]]),
                                  c(1,0.3,1,0.3)))

plot1<-rasterize(plot1, layers='Point', dpi=1000)

plot1

ggsave(file="/Users/gimoro/Desktop/v1_analysis_thesis/
       v1_first_experiment_files/figure-html/genes.svg", 
       plot=plot1, 
       width=2.7, 
       height=2.7)

plot2<-ggplot(cd, 
              aes(x=factor(sample_id,levels=c("downsampled_unmod_1","downsampled_unmod_3",
                                              "downsampled_mod_2","downsampled_mod_4")), 
                  y=total)) +
        geom_violin()+  
  theme_bw()+
  rock_roi_theme_multiple_panels()+
  xlab("")+
  ylab("# transcripts")+ 
  geom_jitter(aes(color=sample_id),
              shape=16, 
              position=position_jitter(0.2),
              alpha=0.3,
              size=0.1)+
  geom_boxplot(width=0.3,
               outlier.size = 0.5,
               alpha=0.5)+
  theme(legend.position="none")+
  scale_color_manual(labels=c("downsampled_unmod_1","downsampled_unmod_3",
                              "downsampled_mod_2","downsampled_mod_4"), 
                     values=alpha(c(myscale[["unmod"]],myscale[["unmod"]],myscale[["rock"]],myscale[["rock"]]),
                                  c(1,0.3,1,0.3)))

plot2<-rasterize(plot2, layers='Point', dpi=1000)

plot2

ggsave(file="/Users/gimoro/Desktop/v1_analysis_thesis/
       v1_first_experiment_files/figure-html/total.svg", 
       plot=plot2, 
       width=2.7, 
       height=2.7)

plot3<-ggplot(cd, aes(x=factor(sample_id,levels=c("downsampled_unmod_1","downsampled_unmod_3",
                                                  "downsampled_mod_2","downsampled_mod_4")), 
                      y=subsets_mouse_mito_percent)) +
        geom_violin()+  
  theme_bw()+
  rock_roi_theme_multiple_panels()+
  geom_jitter(aes(color=sample_id),
              shape=16, 
              position=position_jitter(0.2),
              alpha=0.3,
              size=0.1)+
  xlab("")+
  ylab("mitochondrial transcripts (%)")+
  theme(legend.position="none")+
  scale_color_manual(labels=c("downsampled_unmod_1","downsampled_unmod_3",
                              "downsampled_mod_2","downsampled_mod_4"), 
                     values=alpha(c(myscale[["unmod"]],myscale[["unmod"]],myscale[["rock"]],myscale[["rock"]]),
                                  c(1,0.3,1,0.3)))

plot3<-rasterize(plot3, layers='Point', dpi=1000)

plot3

ggsave(file="/Users/gimoro/Desktop/v1_analysis_thesis/
       v1_first_experiment_files/figure-html/percent_mito.svg", 
       plot=plot3, 
       width=2.7,
       height=2.7)
  
plot4<-ggplot(cd, aes(x = total,
               y = detected,
               colour = subsets_mouse_mito_percent)) +
  geom_point(size=0.5,alpha=0.5,stroke=NA) +
  scale_x_log10() + scale_y_log10() +
  facet_wrap(~factor(sample_id,levels=c("downsampled_unmod_1","downsampled_unmod_3",
                                        "downsampled_mod_2","downsampled_mod_4")))+
  theme_bw()+
  xlab("# transcripts")+
  ylab("# genes")+
  theme_bw()+
  rock_roi_theme_multiple_panels()+
  scale_color_viridis()+
  labs(colour="mt transcripts (%)")

plot4<-rasterize(plot4, layers='Point', dpi=1000)

plot4

ggsave(file="/Users/gimoro/Desktop/v1_analysis_thesis/
       v1_first_experiment_files/figure-html/qc1.svg", 
       plot=plot4, 
       width=2.7, 
       height=2.7)
  
plot5<-ggplot(cd, aes(x = total, 
                      y = subsets_mouse_mito_percent)) +
  geom_point(size=0.5,alpha=0.5,stroke=NA) + 
  scale_x_log10() + 
  scale_y_sqrt() +
  facet_wrap(~factor(sample_id,levels=c("downsampled_unmod_1","downsampled_unmod_3",
                                        "downsampled_mod_2","downsampled_mod_4")))+
  geom_hline(yintercept=c(2,25), colour="orange") +
  geom_vline(xintercept=c(2300), colour="orange") +
  theme_bw()+
  xlab("total number of transcripts")+
  ylab("percent mitochondrial reads")+
  theme_bw()+
  rock_roi_theme_multiple_panels()

plot5<-rasterize(plot5, layers='Point', dpi=1000)

plot5

ggsave(file="/Users/gimoro/Desktop/v1_analysis_thesis/v1_first_experiment_files/
       figure-html/qc2.svg", 
       plot=plot5, 
       width=2.7, 
       height=2.7)
 
```

### Filtering dataset

```{r}

mito <- sce$subsets_mouse_mito_percent
sce_filt <- sce[,sce$total>2300 & mito>2 & mito<25]
sce_filt <- sce_filt[rowSums(assay(sce_filt,1))>2 | 
                    rowData(sce_filt)$gene_type=="capture" ,] 

rm(sce)

```

### Doublet removal

```{r}

sces_adj <- sce_filt
counts(sces_adj) <- assay(sce_filt, 1)
colnames(sces_adj) <- rownames(colData(sces_adj))

sces_nd <- scDblFinder(sces_adj, samples = "sample_id")

table(sces_nd$scDblFinder.class)

sces_nd <- sces_nd[,sces_nd$scDblFinder.class=="singlet"]

#saveRDS(sces_nd,"/Users/gimoro/Desktop/v1_analysis_thesis/data_first_experiment/sces_nd")

rm(sce_filt,sces_nd)

sces_nd<-readRDS(paste0("/Users/gimoro/Desktop/v1_analysis_thesis/",
                           "data_first_experiment/sces_nd"))

table(sces_nd$sample_id)

```

### Correlation dT information of all samples

```{r,out.width = '70%'}

test<-logNormCounts(sces_nd)

summed <- aggregateAcrossCells(test, 
    ids=colData(test)[,c("sample_id")])

summed<-summed[-which(rowSums(assay(summed))==0),]

length(rownames(summed))

logcpm<-as.data.frame(cpm(summed,log=TRUE,prior.count=1))

colnames(logcpm)<-gsub("downsampled_","",colnames(logcpm))

tiff(paste0("/Users/gimoro/Desktop/v1_analysis_thesis/v1_first_experiment_files/",
     "correlation.tiff"), 
     units="in", 
     width=2.7, 
     height=2.7, 
     res=1000)

ggpairs(logcpm,
        diag = list(continuous = "blankDiag"),
        upper = list(continuous = GGally::wrap(ggally_cor, 
                                               stars = F,
                                               size = 4,
                                               col="black")),
        lower = list(continuous = wrap("points", size=0.1)))+
  theme_bw()+
  rock_roi_theme_multiple_panels()+
  theme(axis.line=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank())

dev.off()

```

### Try same thing without eGFP in dataset to see how similar it is --> adding egfp as altExp as it is a confounding factor (at this step as eGFP is part of the total counts and removed afterwards)

```{r,out.width = '70%'}

no_egfp_sce<-sces_nd

altExp(no_egfp_sce)<-SummarizedExperiment(assay(no_egfp_sce[1:8,]))
sub<-no_egfp_sce[9:length(rownames(no_egfp_sce)),]

sub<-logNormCounts(sub)

sub_summed <- aggregateAcrossCells(sub, 
    ids=colData(sub)[,c("sample_id")])

sub_logcpm<-as.data.frame(cpm(sub_summed,log=TRUE,
                              prior.count=1))

colnames(sub_logcpm)<-gsub("downsampled_","",colnames(sub_logcpm))

tiff(paste0("/Users/gimoro/Desktop/v1_analysis_thesis/v1_first_experiment_files/",
     "no_egfp_correlation.tiff"), 
     units="in", 
     width=2.7, 
     height=2.7, 
     res=1000)

ggpairs(sub_logcpm,
        diag = list(continuous = "blankDiag"),
        upper = list(continuous = GGally::wrap(ggally_cor, 
                                               stars = F,
                                               size = 4,
                                               col="black")),
        lower = list(continuous = wrap("points", 
                                       size=0.1)))+
  theme_bw()+
  rock_roi_theme_multiple_panels()+
  theme(axis.line=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank())

dev.off()

```

### Defining captured genes

```{r}

sces_nd$CDS<-colSums(counts(sces_nd)[c("ROI_egfp_ROI_egfp",
                                       paste0("capture_sequence_double_egfp_egfp_",
                                              "capture_sequence_double_egfp_egfp"),
                                       "RoCK_egfp_RoCK_egfp",
                                       "5_to_ROI_egfp_5_to_ROI_egfp",
                                       "3_to_capture_3_to_capture"),])

sces_nd$tx<-colSums(counts(sces_nd)[c("ROI_egfp_ROI_egfp",
                                       paste0("capture_sequence_double_egfp_egfp_",
                                              "capture_sequence_double_egfp_egfp"),
                                       "RoCK_egfp_RoCK_egfp",
                                       "5_to_ROI_egfp_5_to_ROI_egfp",
                                       "3_to_capture_3_to_capture",
                                       "5_UTR_egfp_WPRE_5_UTR_egfp_WPRE",
                                       "3_UTR_egfp_WPRE_3_UTR_egfp_WPRE"),])

```

### % positive cells

```{r,out.width = '70%'}

positive_cells<-data.frame(matrix(nrow=4,ncol=2))

colnames(positive_cells)<-c("CDS","tx")

cd<-colData(sces_nd)

rownames(positive_cells)<-unique(cd$sample_id)

for (i in rownames(positive_cells)){
  positive_cells[i,1]<-length(which(cd[which(cd$sample_id==i),]$CDS>0))/
                       length(rownames(cd[which(cd$sample_id==i),]))*100
  positive_cells[i,2]<-length(which(cd[which(cd$sample_id==i),]$tx>0))/
                       length(rownames(cd[which(cd$sample_id==i),]))*100
}

plottable<-melt(positive_cells)
plottable$sample<-rep(rownames(positive_cells),2)
plottable

plot8<-ggplot(data=plottable,
       aes(x = factor(sample,levels=c("downsampled_unmod_1","downsampled_unmod_3",
                                      "downsampled_mod_2","downsampled_mod_4")),
           y=value,color=variable,fill=variable)) +
  geom_bar(stat="identity",width = 0.8, 
           position = position_dodge(width = 0.9))+
  xlab(" ")+
  ylab("positive cells (%)")+
  theme_bw()+
  labs(color='') +
  ylim(0,100)+
  theme_bw()+
  rock_roi_theme()+
  geom_text(aes(label=format(round(value,2),nsmall=2)), 
            position=position_dodge2(width=1), 
            hjust=-1,
            show.legend=FALSE,
            size=2,angle=90,
            color="black")+
  scale_fill_manual(values=c("CDS"="forestgreen","tx"="white"))+
  scale_color_manual(values=c("black","black"))

plot8

plot8<-rasterize(plot8, layers='Point', dpi=1000)

ggsave(file="/Users/gimoro/Desktop/v1_analysis_thesis/v1_first_experiment_files/
       figure-html/percentage.svg", 
       plot=plot8, 
       width=2.7, 
       height=2.7)
  
```

### Number of UMIs for CDS and tx for each sample

```{r,out.width = '70%'}

cd<-as.data.frame(colData(sces_nd))[,c("CDS","tx","sample_id")]

mean(cd[cd$sample_id=="downsampled_mod_2",]$CDS)
mean(cd[cd$sample_id=="downsampled_mod_2",]$tx)

plottable<-melt(cd)

freq<-as.data.frame(table(plottable$sample_id,plottable$variable,plottable$value))
colnames(freq)<-c("sample_id","region","count","frequency")

plot6<-ggplot(freq, aes(x = count, y = frequency))+
  geom_bar(stat = "identity",
           position = position_dodge2(width = 0.2,  
                                      padding = 0.5),aes(color=region, 
                                                         fill=region))+
  facet_wrap(~factor(sample_id,levels=c("downsampled_unmod_1","downsampled_unmod_3",
                                        "downsampled_mod_2","downsampled_mod_4")))+
  xlab(" ")+
  ylab("frequency")+
  theme_bw()+
  rock_roi_theme_multiple_panels()+
  scale_color_manual(values=c("CDS"="forestgreen",
                              "tx"="grey"))+
  scale_fill_manual(values=c("CDS"="forestgreen",
                             "tx"="grey"))

plot6<-rasterize(plot6, layers='Point', dpi=1000)

plot6

ggsave(file="/Users/gimoro/Desktop/v1_analysis_thesis/v1_first_experiment_files/
       figure-html/total_umis.svg", 
       plot=plot6,
       width=5.4, 
       height=5.4)

```

### Same removing capture region

```{r,out.width = '70%'}

sces_nd$CDS<-colSums(counts(sces_nd)[c("ROI_egfp_ROI_egfp",
                                       "RoCK_egfp_RoCK_egfp",
                                       "5_to_ROI_egfp_5_to_ROI_egfp",
                                       "3_to_capture_3_to_capture"),])

sces_nd$tx<-colSums(counts(sces_nd)[c("ROI_egfp_ROI_egfp",
                                       "RoCK_egfp_RoCK_egfp",
                                       "5_to_ROI_egfp_5_to_ROI_egfp",
                                       "3_to_capture_3_to_capture",
                                       "5_UTR_egfp_WPRE_5_UTR_egfp_WPRE",
                                       "3_UTR_egfp_WPRE_3_UTR_egfp_WPRE"),])


positive_cells<-data.frame(matrix(nrow=4,ncol=2))

colnames(positive_cells)<-c("CDS","tx")

rownames(positive_cells)<-unique(cd$sample_id)

cd<-colData(sces_nd)

for (i in rownames(positive_cells)){
  positive_cells[i,1]<-length(which(cd[which(cd$sample_id==i),]$CDS>0))/
                       length(rownames(cd[which(cd$sample_id==i),]))*100
  positive_cells[i,2]<-length(which(cd[which(cd$sample_id==i),]$tx>0))/
                       length(rownames(cd[which(cd$sample_id==i),]))*100
}

# counts

mean(cd[cd$sample_id=="downsampled_unmod_1",]$CDS)
mean(cd[cd$sample_id=="downsampled_unmod_3",]$CDS)
mean(cd[cd$sample_id=="downsampled_mod_2",]$CDS)
mean(cd[cd$sample_id=="downsampled_mod_4",]$CDS)

```

### Session info

```{r}

sessionInfo()

```


